misc/docerina/src/main/java/org/ballerinalang/docgen/Generator.java

```diff
@@ -326,12 +327,183 @@ public static void addTypeDefinition(TypeDefinitionNode typeDefinition, Module m
+        SyntaxKind syntaxKind = typeDescriptorNode.typeDescriptor().kind();
+
+        switch (syntaxKind) {
+            case RECORD_TYPE_DESC:
+                module.records.add(getRecordTypeModel((RecordTypeDescriptorNode) typeDescriptorNode.typeDescriptor(),
+                        typeName, metaDataNode, semanticModel, module));
+                break;
+            case OBJECT_TYPE_DESC:
+                ObjectTypeDescriptorNode objectTypeDescriptorNode =
+                        (ObjectTypeDescriptorNode) typeDescriptorNode.typeDescriptor();
+                BObjectType bObj = getObjectTypeModel(objectTypeDescriptorNode,
+                        typeName, metaDataNode, semanticModel, module);
+                if (containsToken(objectTypeDescriptorNode.objectTypeQualifiers(), SyntaxKind.SERVICE_KEYWORD)) {
+                    module.serviceTypes.add(bObj);
+                } else {
+                    module.objectTypes.add(bObj);
+                }
+                break;
+            case UNION_TYPE_DESC:
+                Type unionType = Type.fromNode(typeDescriptorNode.typeDescriptor(), semanticModel, module);
+                if (unionType.memberTypes.stream().allMatch(type ->
+                        (type.category != null && type.category.equals("errors")) ||
+                                (type.category != null && type.category.equals("builtin")) &&
+                                        type.name.equals("error"))) {
+                    module.errors.add(new Error(typeName, getDocFromMetadata(metaDataNode),
+                            getDescSectionsDocFromMetaDataList(metaDataNode), isDeprecated(metaDataNode),
+                            Type.fromNode(typeDescriptorNode.typeDescriptor(), semanticModel, module)));
+                } else {
+                    module.unionTypes.add(getUnionTypeModel(typeDescriptorNode.typeDescriptor(),
+                            typeName, metaDataNode, semanticModel, module));
+                }
+                break;
+            case SIMPLE_NAME_REFERENCE, QUALIFIED_NAME_REFERENCE:
+                Type refType = Type.fromNode(typeDescriptorNode.typeDescriptor(), semanticModel, module);
+                if (refType.category.equals("errors")) {
+                    module.errors.add(new Error(typeName, getDocFromMetadata(metaDataNode),
+                            getDescSectionsDocFromMetaDataList(metaDataNode), isDeprecated(metaDataNode), refType));
+                } else {
+                    BType bType = getUnionTypeModel(typeDescriptorNode.typeDescriptor(), typeName,
+                            metaDataNode, semanticModel, module);
+                    bType.isNullable = true;
+                    module.simpleNameReferenceTypes.add(bType);
+                }
+                break;
+            case DISTINCT_TYPE_DESC:
+                if (((DistinctTypeDescriptorNode) (typeDescriptorNode.typeDescriptor())).typeDescriptor().kind()
```

Comment:
Extract `DistinctTypeDescriptorNode` and `kind` to two variables and use in other places.

Reply:
And better to move this switch case to another method.